### DISCUSSION TITLE
Help mocking an extended `window` object.

### DISCUSSION THREAD

**Discussion by mikeymop** [vitest-dev/vitest#2369] [Q&A] [answered]
Hello,

I am attempting to write unit tests for our React application. Our application runs in a kubes cluster and requires the ability to change variables are runtime.

In order to do this, we extend `window` with the global environment variables we wish to use in the application. These are stored under `window.$env`. Part of this object includes the necessary values to authenticate with our IDP.

Because we extend window in order to authenticate our network requests, I do not think we can simply `spyOn` or add the request to an msw restHandler like in the vitest examples.

I saw in the vitest [documentation](https://vitest.dev/guide/mocking.html#globals) that we can mock globals which will reside under `window` and tried using `vi.mockGlobal()` with no luck.

After some searching online I saw some suggestions to use `spyOn` to mock out the property on the window object that is needed.

```ts
const originalWindow = { ...window };
const windowSpy = vi.spyOn(window, '$env', 'get');

  beforeAll(() => {
    windowSpy.mockImplementation(() => ({
      ...originalWindow,
      $env: mockEnv
    }));

  // I've also tried
  vi.stubGlobal('$env', mockEnv);

  });
```
Where `mockEnv` is an imported object containing the necessary values. 

My vite.config.ts looks like so:
```ts
test: {
    cache: false,
    coverage: {
      all: true,
      reporter: ['text', 'json', 'html'],
    },
    globals: true,
    environment: 'jsdom',
    setupFiles: './src/test/setup.ts',
  },
```

However when I run my unit tests I get an error that `window.$env.*` is undefined. I've tried a few different mocking implementations under both `vi` and `msw` with no luck. The sticking point is that we access `window.$env` almost immediately in our `index.tsx` component which serves as our root element for the react runtime.

Ideally, I would like to add these mocks to a `beforeAll()` so that other developers will not have to worry about mocking `window.$env` going forward.

Can any guidance be provided towards achieving this goal?

**Comment by 51fe** [comment]
Hi, '$env' is a property not a method

**Reply by mikeymop** [reply]
Yes, my intention is to use it as a property.
Does this mean that `spyOn()` is the incorrect method to be used to mock `window.$env`?

I feel that I am missing some key understanding because regardless to what I try it's still undefined during the actual test run.

**Reply by 51fe** [reply]
FYR: https://cn.vitest.dev/api/#vi-spyon

**Comment by sheremet-va** [comment]
```ts
 const windowSpy = vi.spyOn(window, '$env', 'get');

  beforeAll(() => {
    windowSpy.mockImplementation(() => ({
      ...originalWindow,
      $env: mockEnv
    }));
```

This is not a "windowSpy", this a spy for `$env` getter, and you need to return only what is accessible on `$env`.

There is also no need to copy window object back and forth. I also don't know what "mockEnv" is in your example.

**Comment by mikeymop** [comment]
I've made some progress.

```ts
  beforeEach(() => {
    vi.stubGlobal('$env', mockEnv);
    console.log('Global env', window.$env);
  });

  test('My Component Mounts', () => {
    expect(window.$env.auth).toBeDefined(); // passes
  });
```

This test passes, however once I try to mount a component things break down.

When the component mounts, it requests data from the network, however I do not get the 'mock network request' step because the window object that I mocked doesn't seem to be maintained when the component begins to render.

```ts
  beforeEach(() => {
    vi.stubGlobal('$env', mockEnv);
    console.log('Global env', window.$env);
  });

  test('My Component Mounts', () => {
    expect(window.$env.auth).toBeDefined(); // previously passed
    render(<MyComponent />); // fails window.$env.auth is undefined
  });
```

Is there any reason my decision to use `jsdom` to mock the browser environment is not recommended? Should I use `happy-dom` instead?

**Comment by mikeymop** [comment, marked as answer]
Hello I have since figured out how to best mock network requests for my needs.

```ts
const fakePostResponse = { ... };
...

describe('My networked component test', () => {
  test('Component Mounts', () => {
    vi.mock('../../apis/BackendClient', () => ({  // this mocks out a node module
      default: vi.fn(), // Mock a default export to make this a valid module
      getPosts: vi.fn(() => Promise.resolve([fakePostResponse, {}])) // mocks the actual func we're using with fake data we defined in this test
    }));

    const ref = React.createRef<HTMLInputElement>();
    const onChangeCallback = vi.fn();

    render(<PostList innerRef={ref} onClick={onClickCallback} />);

    const postList = document.querySelector('#posts');
    expect(postList).toBeDefined();
  });
});
```

When mocking the entire module my function lives in, I no longer break out of the test env into my main codebase.
